middle_end/flambda2/simplify/simplify_apply_expr.ml

```diff
@@ -120,53 +120,45 @@ let rebuild_non_inlined_direct_full_application apply ~use_id ~exn_cont_use_id
   in
   after_rebuild expr uacc
 
-let simplify_direct_full_application ~simplify_expr dacc apply function_decl_opt
+let simplify_direct_full_application ~simplify_expr dacc apply function_decl
     ~callee's_code_id ~result_arity ~down_to_up ~coming_from_indirect =
   let inlined =
-    match function_decl_opt with
-    | None ->
-      (* CR mshinwell: Make sure no other warnings or inlining report decisions
-         get emitted when not rebuilding terms. *)
-      if not (DA.do_not_rebuild_terms dacc)
-      then begin
-        Inlining_report.record_decision
-          (At_call_site
-             (Non_inlinable_function
-                { code_id = Code_id.export callee's_code_id }))
-          ~dbg:(DE.add_inlined_debuginfo' (DA.denv dacc) (Apply.dbg apply));
-        warn_not_inlined_if_needed apply
-          "[@inlined] attribute was not used on this function application (the \
-           optimizer decided not to inline the function given its definition)"
-      end;
-      None
-    | Some function_decl -> (
-      let decision =
-        Call_site_inlining_decision.make_decision dacc ~simplify_expr ~apply
-          ~function_decl ~return_arity:result_arity
-      in
-      let code_id =
-        T.Function_declaration_type.Inlinable.code_id function_decl
-      in
+    (* CR mshinwell: Make sure no other warnings or inlining report decisions
+       get emitted when not rebuilding terms. *)
+    let decision =
+      Call_site_inlining_decision.make_decision dacc ~simplify_expr ~apply
+        ~function_decl ~return_arity:result_arity
+    in
+    if not (DA.do_not_rebuild_terms dacc)
+    then
       Inlining_report.record_decision
         (At_call_site
-           (Inlinable_function { code_id = Code_id.export code_id; decision }))
+           (Known_function
+              { code_id = Code_id.export callee's_code_id; decision }))
```

Comment:
The `code_id` in the decision used to be `T.Function_declaration_type.Inlinable.code_id function_decl`.
Now it is the code_id coming from the type, Those are probably the same. I don't know if this changes anything, or if we should assert that those are the same.

Reply:
Compared to the code ID from `function_decl`, `callee's_code_id` has gone through a meet with the code ID from the call kind of the application. I'm not sure if we can get practical examples where the meet is more precise, but in theory at least it looks like `callee's_code_id` is the one we should report.

Reply:
Yeah, that sounds right.  @chambart do you agree?